### ISSUE TITLE
Make validators compatible with `Quantity` type annotations that preserve units

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#1414]
In #1413, I raised an issue to adopt `astropy.units.Quantity` [Quantity type annotations that preserve units](https://docs.astropy.org/en/stable/units/type_hints.html#preserving-units).  

Before we can resolve #1413, we would need to make the decorators in `plasmapy.utils.decorators.validators` compatible with annotations like:
```Python
import astropy.units as u
from typing import Optional, Union

def f(x: u.Optional[u.Quantity[u.m]]) -> u.Quantity[u.m]):
    ...

def g(x: Union[u.Quantity[u.K], u.Quantity[u.eV]]) -> Optional[u.Quantity[u.K]]:
    ...
```

This might be a good first issue for someone who is familiar with Python but not PlasmaPy.

**Comment by RAJAGOPALAN-GANGADHARAN** [comment]
@namurphy Can I give this a shot?

**Comment by rocco8773** [comment]
@namurphy and I have gone back and forth about this implementation.  I honestly don't like Quantity's type annotation in its current form because it will make the signature line in our code and in our docs really messy.  With that said, I'm still up for supporting it, but only if we still support the older style of typing.  This issue would be additional functionality and not replacement functionality.

To implement this, you are actually going to have to modify `CheckUnits` instead of `ValidateQuantities`, since `ValidateQuantities` is a subclass of `CheckUnits`.  I think you'll only need to touch [these lines of code](https://github.com/PlasmaPy/PlasmaPy/blob/9b4af015c0a8716bff450bfd4fe3d26d88091f7e/plasmapy/utils/decorators/checks.py#L620-L627).

As a side note, @namurphy maybe we should start looking into implementing [Python stub files (`.pyi`)](https://peps.python.org/pep-0484/#stub-files).  I don't know their workings in detail but it might be worth exploring.

@RAJAGOPALAN-GANGADHARAN With all my rambling over, yes have at it!!!

**Comment by namurphy** [comment]
Closed by #2346.
